### ISSUE TITLE
a11y: TextField with maxLength has a11y problems

### ISSUE THREAD

**Issue by andywolff** [flutter/flutter#38940]
Internal: b/140829380

If you have a text field with maxLength, such as

`TextField(maxLength: 3, decoration: InputDecoration(hintText: "hint text test"))`

it behaves poorly in TalkBack.

When I try to navigate to a TextField like this, it focuses only on the 0/3 character count, skipping the input field. The character count reads nicely, announcing "3 characters remaining", but the hintText is not read and it does not indicate that a text field is in focus and can be interacted with. If you double-tap now, it will open the keyboard, but the TalkBack outline is still around the 0/3. I guess this is in some kind of bad initial state.

Now, type some text. When you type the 4th character and exceed the limit, it correctly does not add the character, but it just reads "Showing English (US)(QUERTY) keyboard". I would expect it to provide some indication that the 4th character has been rejected, like "no characters remaining".

Now navigate backwards and then forwards again. Now that the text field has text in it, it reads "Editing, xxx, hint text test, edit box". You can then navigate forwards again to focus on the 3/3, which will properly read "No characters remaining".

If you delete all of the input text, it does not go back into the bad initial state where the screen reader does not see the input field.

**Comment by rymotion** [comment]
You may want to modify the textfield widget and wrap widget used for count with ```mergesemantics``` widget

**Comment by HansMuller** [comment]
CC @gspencergoog - one of the problems described here appears to be with focus traversal. Are we really skipping the input field and landing on the character counter?

**Comment by goderbauer** [comment]
This is the semantics tree for this setup:
```
I/flutter (19606): SemanticsNode#0
I/flutter (19606):  │ Rect.fromLTRB(0.0, 0.0, 1440.0, 2392.0)
I/flutter (19606):  │
I/flutter (19606):  └─SemanticsNode#1
I/flutter (19606):    │ Rect.fromLTRB(0.0, 0.0, 411.4, 683.4) scaled by 3.5x
I/flutter (19606):    │ textDirection: ltr
I/flutter (19606):    │
I/flutter (19606):    └─SemanticsNode#2
I/flutter (19606):      │ Rect.fromLTRB(0.0, 0.0, 411.4, 683.4)
I/flutter (19606):      │ flags: scopesRoute
I/flutter (19606):      │
I/flutter (19606):      ├─SemanticsNode#5
I/flutter (19606):      │ │ Rect.fromLTRB(0.0, 0.0, 411.4, 80.0)
I/flutter (19606):      │ │ thickness: 4.0
I/flutter (19606):      │ │
I/flutter (19606):      │ └─SemanticsNode#6
I/flutter (19606):      │     Rect.fromLTRB(16.0, 40.5, 242.0, 63.5)
I/flutter (19606):      │     flags: isHeader, namesRoute
I/flutter (19606):      │     label: "Flutter Demo Home Page"
I/flutter (19606):      │     textDirection: ltr
I/flutter (19606):      │     elevation: 4.0
I/flutter (19606):      │
I/flutter (19606):      ├─SemanticsNode#3
I/flutter (19606):      │ │ Rect.fromLTRB(0.0, 349.2, 411.4, 414.2)
I/flutter (19606):      │ │ actions: tap
I/flutter (19606):      │ │ flags: isTextField
I/flutter (19606):      │ │ textDirection: ltr
I/flutter (19606):      │ │
I/flutter (19606):      │ └─SemanticsNode#4
I/flutter (19606):      │     Rect.fromLTRB(392.4, 51.0, 411.4, 65.0)
I/flutter (19606):      │     label: "3 characters remaining"
I/flutter (19606):      │     textDirection: ltr
I/flutter (19606):      │
I/flutter (19606):      └─SemanticsNode#7
I/flutter (19606):        │ merge boundary ⛔️
I/flutter (19606):        │ Rect.fromLTRB(0.0, 0.0, 56.0, 56.0) with transform
I/flutter (19606):        │ [1.0,2.4492935982947064e-16,0.0,339.42857142857144;
I/flutter (19606):        │ -2.4492935982947064e-16,1.0,0.0,611.4285714285714;
I/flutter (19606):        │ 0.0,0.0,1.0,0.0; 0.0,0.0,0.0,1.0]
I/flutter (19606):        │ label: "Increment"
I/flutter (19606):        │ textDirection: ltr
I/flutter (19606):        │
I/flutter (19606):        └─SemanticsNode#8
I/flutter (19606):            merged up ⬆️
I/flutter (19606):            Rect.fromLTRB(0.0, 0.0, 56.0, 56.0)
I/flutter (19606):            actions: tap
I/flutter (19606):            flags: isButton, hasEnabledState, isEnabled
I/flutter (19606):            thickness: 6.0
```

Looks like there may be a couple of bugs here
1. The hint text is not included in the semantics tree at all.
2. For some reason android thinks that the Textfield node is not focusable (it works on iOS).
3. There may not be sufficient feedback when a character is not accepted by the textfield (not sure what the right feedback is here).

**Comment by goderbauer** [comment]
1) is actually not an issue. The hint text is only included when the textfield has input focus and is then read out when the textfield has a11y focus. Since it for some reason cannot get a11y focus in this case, the hint text is never read out.

**Comment by goderbauer** [comment]
With 2) maybe the problem is that on Android text field a11y nodes are not allowed to have children?

The following suffers under the same problem:

```dart
        TextField(
          decoration: InputDecoration(
            hintText: "hint text test",
            suffixIcon: IconButton(
              icon: Icon(Icons.add_circle, semanticLabel: 'foo'),
              onPressed: () {},
            ),
          ),
        ),
```

**Comment by gspencergoog** [comment]
Focus traversal and a11y focus traversal are different beasts (and different code), so I doubt that the traversal code is the problem.  I did add the ability to make a field not focusable, but that hasn't been used anywhere yet that I know of (except in my PR that hasn't landed yet).

**Comment by jkurtw** [comment]
b/140829380

**Comment by jkurtw** [comment]
makes several important text fields more confusing for screenreader users, but the required actions are still possible

**Comment by goderbauer** [comment]
I had to roll the change back because it broke a devicelab test that needs updating. Will apply the patch again with an update to that test.
